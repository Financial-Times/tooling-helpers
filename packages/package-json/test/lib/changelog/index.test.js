const createChangelog = require("../../../src/lib/changelog");

let changelog;

beforeEach(() => {
	changelog = createChangelog();
});

describe("createChangelog", () => {
	test("returns object with methods for manipulating changelog", () => {
		expect(changelog).toEqual(expect.any(Object));
		Object.values(changelog).forEach(method => {
			expect(method).toEqual(expect.any(Function));
		});
	});

	test("throws an error if the `event` value is invalid", () => {
		expect(() => {
			changelog.createEntry({
				event: "invalidEvent",
				field: "license"
			});
		}).toThrowErrorMatchingSnapshot();
	});

	test("throws an error if the `field` property is missing", () => {
		expect(() => {
			changelog.createEntry({
				event: "setField"
			});
		}).toThrowErrorMatchingSnapshot();
	});
});

describe("createEntry", () => {
	test("returns a changelog entry object", () => {
		const changelogEntry = changelog.createEntry({
			event: "setField",
			field: "license"
		});
		expect(changelogEntry).toEqual(expect.any(Object));
		expect(changelogEntry).toMatchSnapshot();
	});
});

describe("get", () => {
	test("returns the changelog as an array of entry objects", () => {
		changelog.createEntry({
			event: "setField",
			field: "name",
			previousValue: "previous-name"
		});
		changelog.createEntry({
			event: "requireDependency",
			field: "dependencies",
			pkg: "express",
			version: "4.16.4"
		});
		const changelogEntries = changelog.get();
		expect(changelogEntries).toEqual(expect.any(Array));
		expect(changelogEntries[0]).toEqual(expect.any(Object));
		expect(changelogEntries).toMatchSnapshot();
	});
});

describe("getAsMessages", () => {
	test("returns the changelog as an array of messages", () => {
		changelog.createEntry({
			event: "setField",
			field: "name",
			alreadyExisted: true
		});
		changelog.createEntry({
			event: "requireDependency",
			field: "dependencies",
			pkg: "express",
			version: "4.16.4"
		});
		const changelogEntries = changelog.get.asMessages();
		expect(changelogEntries).toEqual(expect.any(Array));
		expect(changelogEntries[0]).toEqual(expect.any(String));
		expect(changelogEntries).toMatchSnapshot();
	});
});

describe("getLastEntry", () => {
	test("returns object for last entry in the changelog", () => {
		changelog.createEntry({
			event: "setField",
			field: "name",
			previousValue: "previous-name"
		});
		changelog.createEntry({
			event: "requireDependency",
			field: "dependencies",
			pkg: "express",
			version: "4.16.4"
		});
		const lastChangelogEntry = changelog.get.lastEntry();
		expect(lastChangelogEntry).toEqual(expect.any(Object));
		expect(lastChangelogEntry).toMatchSnapshot();
		expect(lastChangelogEntry).toEqual(changelog.get().pop());
	});

	test("returns an empty object if there are no entries in the changelog", () => {
		expect(changelog.get.lastEntry()).toEqual({});
	});
});

describe("getLastEntryAsMessage", () => {
	test("returns last entry in the changelog as a message", () => {
		changelog.createEntry({
			event: "setField",
			field: "name",
			previousValue: "previous-name"
		});
		changelog.createEntry({
			event: "requireDependency",
			field: "dependencies",
			pkg: "express",
			version: "4.16.4"
		});
		const lastChangelogEntry = changelog.get.lastEntryAsMessage();
		expect(lastChangelogEntry).toEqual(expect.any(String));
		expect(lastChangelogEntry).toMatchSnapshot();
		expect(lastChangelogEntry).toEqual(changelog.get.asMessages().pop());
	});

	test("returns an empty string if there are no entries in the changelog", () => {
		expect(changelog.get.lastEntryAsMessage()).toEqual("");
	});
});
