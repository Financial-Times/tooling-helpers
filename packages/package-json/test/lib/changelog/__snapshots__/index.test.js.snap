// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createChangelog throws an error if the \`event\` value is invalid 1`] = `"changelog#createEntry: Entry has invalid \`event\` value: invalidEvent"`;

exports[`createChangelog throws an error if the \`field\` property is missing 1`] = `"changelog#createEntry: Entry is missing the \`field\` property"`;

exports[`createEntry returns a changelog entry object 1`] = `
Object {
  "alreadyExisted": false,
  "event": "setField",
  "field": "license",
  "meta": Object {},
}
`;

exports[`get returns the changelog as an array of entry objects 1`] = `
Array [
  Object {
    "alreadyExisted": false,
    "event": "setField",
    "field": "name",
    "meta": Object {},
    "previousValue": "previous-name",
  },
  Object {
    "alreadyExisted": false,
    "event": "requireDependency",
    "field": "dependencies",
    "meta": Object {
      "pkg": "express",
      "version": "4.16.4",
    },
  },
]
`;

exports[`getAsMessages returns the changelog as an array of messages 1`] = `
Array [
  "Set value for field 'name' (overwrote existing value)",
  "Required package express@4.16.4 in dependencies (new dependency)",
]
`;

exports[`getLastEntry returns object for last entry in the changelog 1`] = `
Object {
  "alreadyExisted": false,
  "event": "requireDependency",
  "field": "dependencies",
  "meta": Object {
    "pkg": "express",
    "version": "4.16.4",
  },
}
`;

exports[`getLastEntryAsMessage returns last entry in the changelog as a message 1`] = `"Required package express@4.16.4 in dependencies (new dependency)"`;
